src/coreclr/jit/llvmlssa.cpp

```diff
@@ -412,26 +407,25 @@ class ShadowStackAllocator
         // shadow frame parameter to filter funclets.
         if (m_llvm->m_unwindFrameLclNum != BAD_VAR_NUM)
         {
-            assignOffset(m_compiler->lvaGetDesc(m_llvm->m_unwindFrameLclNum), TARGET_POINTER_SIZE);
+            assignOffset(m_compiler->lvaGetDesc(m_llvm->m_unwindFrameLclNum));
         }
 
-#ifndef TARGET_64BIT
-        // We assign offsets to the variables that require double alignment first to pack them together.
+        // Assigns offsets such that locals which need to be zeroed come first. This will allow us to zero them all
+        // using a single memset in the prolog.
+        m_prologZeroingOffset = offset;
+
         for (unsigned i = 0; i < shadowFrameLocals.size(); i++)
         {
             LclVarDsc* varDsc = m_compiler->lvaGetDesc(shadowFrameLocals.at(i));
-            if (m_llvm->isShadowFrameLocal(varDsc))
+            if (m_llvm->isShadowFrameLocal(varDsc) || !varDsc->lvMustInit)
```

Comment:
There's something clever here that I dont see.  How is this shuffling the entries so that the `lvMustInit` come first?

Reply:
or rather I see how it moves those first, but not how the others are not shifted out of the "zero init" space.

Reply:
The algorithm groups locals into 3 clusters in 3 loops:
```
[cluster 1] [virtual unwind frame]
[cluster 2] [all lvMustInit locals]
[cluster 3] [all other locals]
```
On each subsequent loop, locals from the previous clusters have already been assigned an offset, and so are skipped.